xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,793 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
 func (bb) Name() string {
 	return Name
 }
+
+// scUpdate wraps a subConn update to be sent to the child balancer.
+type scUpdate struct {
+	scw   *subConnWrapper
+	state balancer.SubConnState
+}
+
+type ejectedUpdate struct {
+	scw     *subConnWrapper
+	ejected bool // true for ejected, false for unejected
+}
+
+type outlierDetectionBalancer struct {
+	numAddrsEjected int // For fast calculations of percentage of addrs ejected
+
+	childState       balancer.State
+	recentPickerNoop bool
+
+	closed *grpcsync.Event
+	cc     balancer.ClientConn
+	bOpts  balancer.BuildOptions
+	logger *grpclog.PrefixLogger
+
+	// childMu protects child and also updates to the child (to uphold the
+	// balancer.Balancer API guarantee of synchronous calls). It also protects
+	// against run() reading that the child is not nil for SubConn updates, and
+	// then UpdateClientConnState or Close writing to the the child.
+	childMu sync.Mutex
+	child   balancer.Balancer
+
+	// mu guards access to a lot of the core LB Policy State. It also prevents
+	// intersplicing certain operations.
+	//
+	// ex 1: interval timer goes off, outlier detection algorithm starts running
+	// based on knobs in cfg. in the middle of running the algorithm, a
+	// ClientConn update comes in and writes to cfg. This causes undefined
+	// behavior for the interval timer algorithm.
+	//
+	// ex 2: Updating the addrs map from UpdateAddresses in the middle of
+	// running the interval timer algorithm which uses addrs heavily. This will
+	// cause undefined behavior for the interval timer algorithm.
```

Comment:
Is the comment saying that these undefined behaviors are eliminated because of using this mutex, or it is saying that we still have those undefined behaviors? If the former, please clarify. If the latter, please clarify why this undefined behavior will not affect correctness.

Reply:
This is still a cause for some confusion to me.

The comment is of the following form now:
- mu guards access to blah. It helps eliminate some undefined behaviors:
- example 1: you go over the example, and say "this causes undefined behavior"
- same for example 2

Instead I feel it would be better if:
- mu guards access to the following fields. 
- It helps to synchronize actions taken as a result of the following events: config updates, firing of interval timer, address updates from the child policy etc (are there more?)
- For example, when we receive a config update in the middle of the interval timer algorithm <<< this is how we synchronize >>>
- For example, when we receive an address update in the middle of the interval timer algorithm <<< this is how we synchronize >>>

What do you think?

Reply:
Rewrote. Let me know what you think.